chat.go

```diff
@@ -0,0 +1,124 @@
+// Package chat implements a chat server & client
+package chat
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Server represents a chat server
+type Server struct {
+	Logger         logrus.StdLogger // with standard logger can be extended with logrus
+	WelcomeMessage string
+	listener       net.Listener
+	clients        map[*net.Conn]bool
+	register       chan *net.Conn
+	unregister     chan *net.Conn
+	broadcast      chan string
+}
+
+// Logger enables a customization of the log function
+type Logger interface {
+	Println(v ...interface{})
+}
+
+// Run implements the logic handling connections
+func (s *Server) Run() {
+
+	go s.listen()
+
+	for {
+		select {
+		case conn := <-s.register:
+			s.clients[conn] = true
+			s.Logger.Println("client connection registered")
+			go s.handle(conn)
+		case m := <-s.broadcast:
+			for conn := range s.clients {
+				_, err := fmt.Fprint(*conn, m)
+				if err != nil {
+					s.Logger.Println(fmt.Sprintf("sending message failed: %v", err))
+				}
+			}
+		case conn := <-s.unregister:
+			delete(s.clients, conn)
+			s.Logger.Println("client connection unregistered")
+		}
+	}
+}
+
+func (s *Server) listen() {
+	for {
+		conn, err := s.listener.Accept()
+		if err != nil {
+			s.Logger.Println(fmt.Sprintf("connection failed: %v", err))
+			return
+		}
+		s.register <- &conn
+	}
+}
+
+func (s *Server) handle(conn *net.Conn) {
+	_, err := (*conn).Write([]byte(s.WelcomeMessage + "\n"))
```

Comment:
If we're reading and writing regular messages as strings, we should do the same here, and print it to the conn with `fmt.Fprint`, I think.